circadian/lights.py

```diff
@@ -4,157 +4,400 @@
 __all__ = ['Light', 'make_pulse', 'get_pulse']
 
 # %% ../nbs/01_lights.ipynb 3
-from matplotlib.pyplot import step
 import numpy as np
-from numpy.core.fromnumeric import repeat
 import pylab as plt
 from typing import Optional
+from matplotlib.pyplot import step
+from fastcore.basics import patch_to
+from numpy.core.fromnumeric import repeat
 
-# %% ../nbs/01_lights.ipynb 4
+# %% ../nbs/01_lights.ipynb 5
 class Light:
-
+    "Helper class for creating light schedules"
     def __init__(self, 
                  light: callable, # light function that takes in a time value and returns a float, if a float is passed, then the light function is a constant set to that lux value 
-                 start_time: float = 0.0, # when the light function starts in hours
-                 duration: float = 1.0, # duration of the light function in hours
-                 default_value=0.0 # the default value of the light function when it is not active
+                 start_time: float=0.0, # when the light function starts in hours
+                 duration: float=1.0, # duration of the light function in hours
+                 default_value: float=0.0 # the default value of the light function outside of its duration
                  ):
-        if type(light) is float:
-            self._func = np.vectorize(lambda t: light, otypes=[float])
+        # light type checking
+        light_input_err_msg = "`light` should be a nonnegative `float`, or a callable with a single `float` parameter which returns a nonnegative `float`"
+        if callable(light):
+            light_fn = np.vectorize(light, otypes=[float])
+            output = light_fn(0.0)
+            try:
+                float(output)
+            except:
+                # catches when the function created from light does not return values that can be cast to float
+                raise ValueError(light_input_err_msg)
         else:
-            self._func = np.vectorize(light, otypes=[float])
-        self.default_value = default_value
-        self.start_time = start_time
+            try:
+                light = float(light)
+                light_fn = np.vectorize(lambda t: light, otypes=[float])
+            except:
+                # catches when the light function can't be converted to a float
+                raise ValueError(light_input_err_msg)
+
+        # start_time type checking
+        start_time_err_msg = "`start_time` should be a `float`"
```

Comment:
These are good checks that we should keep, but I normally think of unit tests as something where we assert a property is true in a separate file; e.g. `tests/test_lights.py`. In particular, I'd love to test the following in unit tests:

1) That the model can entrain to a 24 hr RegularLight from random ICs after X repetitions and give the same value at `t` and `t + 24`
2) That we can create various light profiles and the assert to check the values of the lights at different times are what we'd expect (e.g. construct shift work light schedule, assert light at some points, and assert darkness at others)
3) That the new add function works as expected; e.g. add two lights and then assert the values we expect are there. 

How's this sound to you?

Reply:
With respect to test 1. I wanted to add it once I update the code on both the light and models modules because that test will use functions from both. Does it make sense to delay adding that one until I have both modules updated?